### DISCUSSION TITLE
Simplify update process

### DISCUSSION THREAD

**Discussion by foal** [kubeapps/kubeapps#3568] [Ideas]
Hi, 

Right now the complete update process (INHO) to the new chart release contains too many manuals steps:

1. Select an application to update.
2. Copy current values content to the clipboard 
3. Click restore defaults (it loads original values from the new version of Helm chart)
4. Copy new values to the clipboard 
5. Use the clipboard compare tool (e.g. Beyond Compare) to merge changes from the current state into new values (2-way merge)
6. copy/paste updated new values back to the kubeapps
7. wait while kubeapps process new values 
8. check changes on the Changes tab
9. apply the update

For more complex update is possible to take original values of the current version as the base for 3-way merge, for example:

Installed package version: 5
Update to package version: 7

Files for 3-way merge:
Base: **original values.yaml from package version: 5**
Left: your changes - **current values based on package version: 5** (e.g. you enable ingress)
Right: version 7 changes - **original values.yaml from package version: 7** 
Result: **custom (e.g. with enabled ingress) values based on package version: 7**

Best,
Stanislav

**Comment by absoludity** [comment]
Hi there @dstefka . It's not intended to require those manual steps. I could be missing something, but I think that either:

-  you may not have realised that Kubeapps is (trying to) do that merge for you, including your custom changes when you upgrade with the new charts values, rather than overwriting them with the new charts values. Or,
- in the case you have, Kubeapps' merge of your custom values with the new chart values is failing? Let me take a look at your steps with a specific example, or
- you're expecting kubeapps to merge non-custom changes from previous charts to the new one (not sure).

Let me start by installing an older version of the Apache chart, version 7.6.0, but with a custom value: I've explicitly changed the `livenessprobe.failureThreshold` from 6 to 600, so that before installing, I see the following in the changes tab:

![apache-test-1](https://user-images.githubusercontent.com/497518/136716996-69e3b5ae-eca5-4107-b205-5b8e21e63d73.png)

and I install it with those values. One thing to note: before having installed a package, the changes tab shows the difference between your values and the packaged version. This is different when upgrading, as we'll see below.

Once it's installed, I come back to the app and click update, and then ensure I choose the latest version, 8.8.4. If I look at the changes tab, I will see a lot of changes because, as the app is already installed the changes tab is now showing the diff between what is going to be installed and what is currently installed (not the diff between what is going to be installed and the current package). So all the changes that I see are actually changes from the new chart versions' values.yaml, but if I check closely, I see that my custom value is still set (and will persisted by the upgrade). See the `failureThreshold` value below:

![apache-test-2](https://user-images.githubusercontent.com/497518/136718434-a4077cf3-43e4-4105-b4c2-ad3d0d42012a.png)

Is that different to what you expected, or are you seeing something different to that? If so, can you run me through a simple example using (for example) the apache chart as I've done, with screenshot examples?

Thanks!

**Reply by absoludity** [reply]
Or another option: I notice that Antonio documents (and fixes) an issue for a specific situation where the changes aren't preserved at #3540 . Take a look at his demo video on the PR to see if that's the issue you're hitting?

**Reply by foal** [reply]
Yes, right after I wrote the post I have updated the Prometheus chart, and it behaves as you described above. Cool!

But before I have an issue with the ExternalDNS chart - I had changed the image settings in the old (due to k8s 1.22) chart, the new chart also has a new image but the "merged" result contains my "custom" image. With the current interface it is only one solution - try to preserve all "custom/user" settings, but would be better if:
-  interface shows the summary of all custom settings, not only the current diff between releases (now it can be simulated by clicking on the "Restore defaults" button), 
- on **update with the change chart version** shows one more tab with the difference between original charts, and warns about conflicts.

Thanks!

**Reply by antgamdia** [reply]
Yep, right now we are not displaying these changes (you have to restore default as you mentioned). I guess it shouldn't be that difficult adding another diff view (or just a toggle button) to check the changes against the original chart values. Feel free to file an issue so that we can plan it in our iterations... or even better, feel free to send us a PR :)   I'm happy to help or assist you.
